include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -0,0 +1,616 @@
+  }
+
+  void deallocate_no_sync(void *ptr, size_t size, size_t alignment) {
+    deallocate_impl(ptr, size, alignment, false);
+  }
+
+  struct Stat {
+    int allocated_blocks;
+    int peak_allocated_blocks;
+    size_t block_size;
+    size_t allocated_va;
+    size_t curr_allocated;
+    size_t peak_allocated;
+    size_t curr_free;
+    size_t curr_allocations;
+    size_t peak_allocations;
+    size_t total_allocations;
+    size_t total_deallocations;
+    size_t total_unmaps;
+  };
+
+  Stat get_stat() const {
+    lock_guard pool_guard(pool_lock_);
+    return stat_;
+  }
+
+  void clear_stat() {
+    lock_guard pool_guard(pool_lock_);
+    stat_ = {};
+  }
+
+ private:
+  void purge() {
+    lock_guard pool_guard(pool_lock_);
+    mem_lock_guard mem_guard(mem_lock_);
+    for (auto &r : va_regions_)
+      r.purge();
+    va_ranges_.clear();
+  }
+
+  void configure() {
+    if (device_ordinal_ < 0) {
+      CUDA_CALL(cudaGetDevice(&device_ordinal_));
+    }
+    if (block_size_ == 0) {
+      CUdevice device;
+      CUDA_CALL(cuDeviceGet(&device, device_ordinal_));
+      CUDA_CALL(cuDeviceTotalMem(&total_mem_, device));
+      size_t grain = cuvm::GetAddressGranularity();
+      block_size_ = std::max(grain,  // at least grain, for correctness
+                             std::min<size_t>(next_pow2(total_mem_ >> 8),  // capacity-dependent...
+                                              64 << 20));  // ...but capped at 64 MiB
+    }
+    stat_.block_size = block_size_;
+  }
+
+  struct va_region {
+    va_region(cuvm::CUAddressRange range, size_t block_size)
+    : address_range(range), block_size(block_size) {
+      size_t size = address_range.size();
+      size_t blocks_in_range = size / block_size;
+      mapping.resize(blocks_in_range);
+      mapped.resize(blocks_in_range, false);
+      available.resize(blocks_in_range, false);
+    }
+
+    va_region(va_region &&) = default;
+    va_region &operator=(va_region &&) = default;
+
+    ~va_region() {
+      purge();
+    }
+
+    int num_blocks() const {
+      return mapping.size();
+    }
+
+    void purge() {
+      CUdeviceptr ptr = address_range.ptr();
+      for (size_t i = 0; i < mapping.size(); i++, ptr += block_size) {
+        auto &h = mapping[i];
+        if (h) {
+          cuvm::Unmap(ptr, block_size);
+          CUDA_DTOR_CALL(cuMemRelease(h));
+          h = {};
+        }
+      }
+      mapped.fill(false);
+      available.fill(false);
+    }
+
+    CUdeviceptr block_dptr(ptrdiff_t block_index) const noexcept {
+      return address_range.ptr() + block_index * block_size;
+    }
+
+    template <typename T = void>
+    T *block_ptr(ptrdiff_t block_index) const noexcept {
+      return reinterpret_cast<T*>(block_dptr(block_index));
+    }
+
+    void map_block(int block_idx, cuvm::CUMem mem) {
+      assert(mem.size() == block_size);
+      cuvm::Map(block_dptr(block_idx), mem);
+      mapping[block_idx] = mem.release().first;
+      mapped[block_idx] = true;
```

Comment:
Because searching for a mapped / unmapped block is 64x way faster in a bit vector than in a vector of pointers. Also, cache footprint is much smaller. And in real life, this is going to be the most frequent case - that the memory is already mapped.